builtin/logical/database/path_rotate_credentials.go

```diff
@@ -78,22 +78,27 @@ func (b *databaseBackend) pathRotateRootCredentialsUpdate() framework.OperationF
 			return nil, err
 		}
 
-		// Take out the backend lock since we are swapping out the connection
-		b.Lock()
-		defer b.Unlock()
-
 		// Take the write lock on the instance
 		dbi.Lock()
-		defer dbi.Unlock()
-
+		unlocked := false
+		unlock := func() {
+			if !unlocked {
+				// set to true before unlocking to avoid race condition
+				unlocked = true
+				dbi.Unlock()
+			}
+		}
+		defer unlock()
 		defer func() {
 			// Close the plugin
 			dbi.closed = true
 			if err := dbi.database.Close(); err != nil {
 				b.Logger().Error("error closing the database plugin connection", "err", err)
 			}
 			// Even on error, still remove the connection
-			delete(b.connections, name)
+			// be careful with a deadlock here
```

Comment:
As a first time reader of this code, the deadlock wasn't immediately obvious to me. Could we spell it out a bit more?

Reply:
:+1: